mailserver/limiter_test.go

```diff
@@ -0,0 +1,95 @@
+package mailserver
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/suite"
+)
+
+func TestLimiterSuite(t *testing.T) {
+	suite.Run(t, new(LimiterSuite))
+}
+
+type LimiterSuite struct {
+	suite.Suite
+}
+
+func (s *LimiterSuite) TestIsAllowed() {
+	peerID := "peerID"
+	testCases := []struct {
+		t               time.Duration
+		shouldBeAllowed bool
+		db              func() map[string]time.Time
+		errMsg          string
+	}{
+		{
+			t:               time.Duration(5) * time.Millisecond,
+			shouldBeAllowed: true,
+			db: func() map[string]time.Time {
+				return make(map[string]time.Time)
+			},
+			errMsg: "Expected limiter not to allow with empty db",
+		},
+		{
+			t:               time.Duration(5) * time.Millisecond,
+			shouldBeAllowed: true,
+			db: func() map[string]time.Time {
+				db := make(map[string]time.Time)
+				db[peerID] = time.Now().Add(time.Duration(-10) * time.Millisecond)
+				return db
+			},
+			errMsg: "Expected limiter to allow with peer on its db",
+		},
+		{
+			t:               time.Duration(5) * time.Millisecond,
+			shouldBeAllowed: false,
+			db: func() map[string]time.Time {
+				db := make(map[string]time.Time)
+				db[peerID] = time.Now().Add(time.Duration(-1) * time.Millisecond)
+				return db
+			},
+			errMsg: "Expected limiter to not allow with peer on its db",
+		},
+	}
+
+	for _, tc := range testCases {
+		l := newLimiter(tc.t)
+		l.db = tc.db()
+		s.Equal(tc.shouldBeAllowed, l.isAllowed(peerID), tc.errMsg)
+	}
+}
+
+func (s *LimiterSuite) TestRemoveExpiredRateLimits() {
+	peer := "peer"
+	l := newLimiter(time.Duration(5) * time.Second)
+	for i := 0; i < 10; i++ {
+		peerID := fmt.Sprintf("%s%d", peer, i)
+		l.db[peerID] = time.Now().Add(time.Duration(i*(-2)) * time.Second)
```

Comment:
Not sure about this kind of white box testing. Maybe seeing it too strict as I already got troubles calling private methods/functions in tests. 😁

Reply:
If I avoid accessing private `db` here i'll should be using and trusting `limiter.Add` method which is out of the scope for this specific test. 

Another approach could be allowing defining database on object initialization, but maybe that's adding extra logic to the tested object just for testing, so, its neither good approach .